File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int answer = 0;

int calculate(string current, vector<int> numbers) {
    // +, - 계산하기
    int result = 0;
    for(int i = 0; i < numbers.size(); i++) {
        // + 이면
        if(current[i] == '+') {
            result += numbers[i];  
        } else { // - 이면
            result -= numbers[i];
        }
    }
    return result;
}

// 재귀
void recur(string current, int remaining, int target, vector<int> numbers) {
    int result = 0;
    // 다 뽑으면 멈추기, target과 같으면 answer값 +1
    if(remaining == 0) {
        result = calculate(current, numbers);
        if(result == target) {
            answer++;
            return;    
        } else {
            return;
        }
    }
    
    // 재귀 파트
    // + 뽑기
    recur(current + "+", remaining - 1, target, numbers);
    // - 뽑기
    recur(current + "-", remaining - 1, target, numbers);
}

int solution(vector<int> numbers, int target) {
    recur("", numbers.size(), target, numbers); // length()는 string 전용. vector는 size()를 사용해야 한다.
    return answer;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
# numbers 순서 고정, 각 수들을 더하거나 빼서 타겟 넘버 만들어야 한다.
# 타겟 넘버를 만드는 방법의 수 return하기

# 재귀로 모든 경우의 수 끝에 나온 수가 타겟 넘버와 일치하면 answer += 1

def recur(numbers, target, cur_idx, cur_sum):
    # 멈추는 조건: 타겟 넘버와 일치하면 1, 아니면 0 반환
    if cur_idx == len(numbers):
        return 1 if cur_sum == target else 0
    
    # 재귀 파트: (현재가 +일때의 일치 횟수) + (현재가 -일때의 일치 횟수) 반환
    return (
        recur(numbers, target, cur_idx + 1, cur_sum + numbers[cur_idx])
        + recur(numbers, target, cur_idx + 1, cur_sum - numbers[cur_idx])
    )
    

def solution(numbers, target):
    return recur(numbers, target, 0, 0)